### ISSUE TITLE
Spykbean does not work on proxied bean with @Transactional

### ISSUE THREAD

**Issue by phillipuniverse** [Ninja-Squad/springmockk#63]
A very old bug [in Spring Boot](https://github.com/spring-projects/spring-boot/issues/7033#issuecomment-880929037) was fixed recently by [this commit](https://github.com/spring-projects/spring-boot/commit/c8c784bd5ca86faaaecdf2371aa35cf98c62efc5) and I just ran into the same type of bug today with springmockk. It appears that there is some drift potential between the Spring `SpyDefinition` and this project's `SpykDefinition`. I believe the same type of fix will need to be applied to `SpykDefinition`.

Here's the bean that is being tested (some psuedo-code):

```kotlin
@Service
class BulkService(
    private val bulkRepository: BulkRepository
) {

    @Transactional
    fun create(req: BulkRequest, tenantId: String): BulkResponse {
        var bulkOperation = processRequest(bulkOperationRequest, companyId, userId)
        bulkOperation = bulkRepository.save(bulkOperation)
        return buildResponse(bulkOperation)
    }

    internal fun processRequest(request: BulkRequest, tenant: String, userId: String?): BulkOperation {
        ...
    }
}
```

And here's what my test class looks like (again, some psuedo-code):

```kotlin

@SpringBootTest
@TestMethodOrder(MethodOrderer.OrderAnnotation::class)
class DummyTest {

    @SpykBean
    private lateinit var bulkService: BulkService

    @Test
    @Order(1)
    fun mockedMethodInvocation() {
        every { bulkOperationService.processRequest(any(), any()) } returns BulkOperation(...)

        val response = bulkOperationService.create(BulkRequest(...), "abc123")
        // assert response is the mocked version
    }

    @Test
    @Order(2)
    fun realMethodInvocation() {
        val req = BulkRequest()
        val resp = bulkOperationService.create(req, "sometenant")
        assertThat(resp).isNotNull
        // verify this was actually saved
    }
```

When I run this and the mocked method goes first, I get this exception when the `realMethodInvocation()` method runs:

```console

-1
java.lang.ArrayIndexOutOfBoundsException: -1
	at io.mockk.impl.verify.LCSMatchingAlgo.lcs(LCSMatchingAlgo.kt:52)
	at io.mockk.impl.verify.VerificationHelpers.reportCalls$default(VerificationHelpers.kt:61)
	at io.mockk.impl.verify.SequenceCallVerifier$verify$1.invoke(SequenceCallVerifier.kt:27)
	at io.mockk.impl.verify.SequenceCallVerifier$verify$1.invoke(SequenceCallVerifier.kt:12)
	at io.mockk.impl.recording.CommonCallRecorder.safeExec(CommonCallRecorder.kt:72)
	at io.mockk.impl.log.SafeToString.exec(SafeToString.kt:10)
	at io.mockk.impl.verify.SequenceCallVerifier.verify(SequenceCallVerifier.kt:26)
	at io.mockk.impl.recording.states.VerifyingState$recordingDone$outcome$1.invoke(VerifyingState.kt:31)
	at io.mockk.impl.recording.states.VerifyingState$recordingDone$outcome$1.invoke(VerifyingState.kt:11)
	at io.mockk.impl.recording.CommonCallRecorder.safeExec(CommonCallRecorder.kt:72)
	at io.mockk.impl.recording.states.VerifyingState.recordingDone(VerifyingState.kt:30)
	at io.mockk.impl.recording.CommonCallRecorder.done(CommonCallRecorder.kt:47)
	at io.mockk.impl.eval.RecordedBlockEvaluator.record(RecordedBlockEvaluator.kt:60)
	at io.mockk.impl.eval.VerifyBlockEvaluator.verify(VerifyBlockEvaluator.kt:30)
	at io.mockk.MockKDsl.internalVerify(API.kt:118)
	at io.mockk.MockKKt.verify(MockK.kt:149)
	at io.mockk.MockKKt.verify$default(MockK.kt:146)
	at com.mycompany.DummyTest.realMethodInvocation(DummyTest.kt:157)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:688)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:210)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:206)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:65)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.stop(TestWorker.java:135)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.lang.Thread.run(Thread.java:748)
```

If I use `@SpyBean` on the `BulkService` property and use Mockito everything works fine, e.g.:

```kotlin
object MockitoHelper {
    fun <T> anyObject(): T {
        Mockito.any<T>()
        return uninitialized()
    }
    @Suppress("UNCHECKED_CAST")
    fun <T> uninitialized(): T =  null as T
}

...

    @Test
    @Order(1)
    fun mockedMethodInvocation() {
        Mockito.doReturn(BulkOperation(...)).
           `when`(bulkOperationService).processRequest(MockitoHelper.anyObject(), MockitoHelper.anyObject())
```

I'm working on a better test that illustrates this behavior in isolation but wanted to get this issue created while it was fresh.

**Comment by phillipuniverse** [comment]
I added a super trimmed down project at https://github.com/phillipuniverse/spring-mockk-transactional the exhibits the behavior. The issue there is a little different but the underlying problem is still probably the same -- the spy is not reset between test methods while `@SpyBean` works fine.

**Comment by jnizet** [comment]
Thanks for the report. I'll have a look, but not immediately. I should have a bit more time next week.

**Comment by jnizet** [comment]
Hi @phillipuniverse.

I've had a look at your reproduction project, and played a little bit with it.
When using Mockito, your tests pass with version 2.5.3 of Spring Boot, but also pass with version 2.5.0 and 2.4.0. So that means the issue you have with MockK is not the issue that was fixed by a recent commit of Spring Boot, but rather by something older. And my guess is that you're hitting the known limitation of SpringMockK described in the [README](https://github.com/Ninja-Squad/springmockk#limitations). I still have no idea of how to fix that in SpringMockK, so if you have an idea, please send a PR.

BTW, I tried to report [the commit](https://github.com/spring-projects/spring-boot/commit/c8c784bd5ca86faaaecdf2371aa35cf98c62efc5) that is missing in SpringMockK, but 
1. It would probably not fix your issue (see above)
2. the test that has been added in Spring Boot in that commit does not fail with SpringMockK, so maybe SpringMockK doesn't have the issue fixed by this Spring Boot commit
3. I can't use the same strategy as in Spring Boot, because AFAIK, MockK doesn't have the notion of a default answer, and Spring Boot's trick is to create a mock with a default answers that delegates to an instance.

**Comment by phillipuniverse** [comment]
Argh -- this does indeed look like the same issue, thanks for the care in double-checking. I'll go ahead and close this one, maybe it will help someone else in Googling.

> I can't use the same strategy as in Spring Boot, because AFAIK, MockK doesn't have the notion of a default answer, and Spring Boot's trick is to create a mock with a default answers that delegates to an instance.

I took a look and it's certainly not straightforward or clear how to do this. The best I found was tracing through [the spyk invocation](https://github.com/mockk/mockk/blob/c726039b573e3bd7d1d03e441ff1a4759311b22f/mockk/common/src/main/kotlin/io/mockk/impl/instantiation/AbstractMockFactory.kt#L68-L96) and this `SpyKStub `. It _seems_ like it would be easy to modify this to delegate to `objToCopy` or be able to initialize that stub with the `objToCopy` that was passed in (being the real object spied on).

I'll keep my eye out and maybe in a debugging session something will become clear with how to fix. But in any case, as you said it will almost certainly be an added feature in MockK that this library will need to take advantage of.

Also, thanks so much for this library it's great a great wrapper!! A lifesaver for the semantic MockK DSL with Spring!
